enterprise/server/workflow/service/service.go

```diff
@@ -794,6 +798,153 @@ func (ws *workflowService) waitForWorkflowInvocationCreated(ctx context.Context,
+	q.AddWhereClause("role = ?", "CI_RUNNER")
+	q.AddWhereClause("group_id = ?", authenticatedUser.GetGroupID())
+	q.AddWhereClause("updated_at_usec > ?", timeLimitMicros)
+	q.SetOrderBy("created_at_usec", false)
+	q.SetLimit(30)
+
+	return q, nil
+}
+
+func (ws *workflowService) GetWorkflowHistory(ctx context.Context, req *wfpb.GetWorkflowHistoryRequest) (*wfpb.GetWorkflowHistoryResponse, error) {
+	if ws.env.GetDBHandle() == nil || ws.env.GetOLAPDBHandle() == nil {
+		return nil, status.FailedPreconditionError("database not configured")
+	}
+
+	repos := req.GetRepoUrls()
+	authenticatedUser, err := ws.env.GetAuthenticator().AuthenticatedUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// Only fetch workflow data for stuff that has run in the last 7 days.
+	timeLimitMicros := time.Now().Add(-time.Duration(7*24) * time.Hour).UnixMicro()
+
+	// Find the names of all of the actions for each repo.
+	q := query_builder.NewQuery(`SELECT repo_url,pattern,count(1) AS total_runs, countIf(success) AS successful_runs, toInt64(avg(duration_usec)) AS average_duration FROM Invocations`)
+	q.AddWhereClause("repo_url IN ?", repos)
+	q.AddWhereClause("role = ?", "CI_RUNNER")
+	q.AddWhereClause("group_id = ?", authenticatedUser.GetGroupID())
+	q.AddWhereClause("invocation_status = ?", int(inspb.InvocationStatus_COMPLETE_INVOCATION_STATUS))
+	q.AddWhereClause("updated_at_usec > ?", timeLimitMicros)
+	q.SetGroupBy("repo_url,pattern")
+	// XXX: No permissions check here; is this ok?
+
+	type actionQueryOut struct {
+		RepoUrl         string
+		Pattern         string
+		TotalRuns       int64
+		SuccessfulRuns  int64
+		AverageDuration int64
+	}
+
+	qStr, qArgs := q.Build()
+	rows, err := ws.env.GetOLAPDBHandle().RawWithOptions(ctx, clickhouse.Opts().WithQueryName("query_find_workflow_actions"), qStr, qArgs...).Rows()
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	actionHistoryQArgs := make([]interface{}, 0)
+	actionHistoryQStrs := make([]string, 0)
+	workflows := make(map[string]map[string]*wfpb.ActionHistory)
+	for rows.Next() {
+		row := &actionQueryOut{}
+		if err := ws.env.GetOLAPDBHandle().DB(ctx).ScanRows(rows, &row); err != nil {
+			return nil, err
+		}
+		q, err := ws.buildActionHistoryQuery(ctx, row.RepoUrl, row.Pattern, timeLimitMicros)
+		if err != nil {
+			return nil, err
+		}
+		actionQStr, actionQArgs := q.Build()
+		actionHistoryQArgs = append(actionHistoryQArgs, actionQArgs...)
+		actionHistoryQStrs = append(actionHistoryQStrs, "("+actionQStr+")")
+		summary := &wfpb.ActionHistory_ActionHistorySummary{
+			TotalRuns:       row.TotalRuns,
+			SuccessfulRuns:  row.SuccessfulRuns,
+			AverageDuration: durationpb.New(time.Duration(row.AverageDuration) * time.Microsecond),
+		}
+		if _, ok := workflows[row.RepoUrl]; !ok {
+			workflows[row.RepoUrl] = make(map[string]*wfpb.ActionHistory)
+		}
+		workflows[row.RepoUrl][row.Pattern] = &wfpb.ActionHistory{ActionName: row.Pattern, Summary: summary}
+	}
```

Comment:
uber nit: check `rows.Err()` here - not sure if it happens in practice but the docs for `rows.Next()` say that it can fail and `rows.Err()` will contain the failure

Reply:
done here and for the history one below.